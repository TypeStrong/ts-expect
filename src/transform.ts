import * as ts from "typescript";

/**
 * Strip TypeScript expectations from runtime code.
 */
export default function(): ts.TransformerFactory<ts.SourceFile> {
  function visitor(
    context: ts.TransformationContext,
    sourceFile: ts.SourceFile
  ): ts.Visitor {
    let keywords = new Set();

    return function visit(node: ts.Node): ts.VisitResult<ts.Node> {
      if (ts.isImportDeclaration(node)) {
        const importName = node.moduleSpecifier.getText().slice(1, -1);

        if (importName === "ts-expect" && node.importClause) {
          const { namedBindings } = node.importClause;

          if (namedBindings) {
            ts.forEachChild(namedBindings, x => keywords.add(x.getText()));
          }

          return node; // Let minifier handle this.
        }
      }

      if (ts.isFunctionLike(node)) {
        const oldKeywords = new Set(keywords)

        // Remove shadowed keywords.
        node.parameters
          .map(x => x.name.getText())
          .filter(x => keywords.has(x))
          .forEach(x => keywords.delete(x));

        const result = ts.visitEachChild(node, visit, context);
        keywords = oldKeywords; // Restore keywords.
        return result;
      }

      if (ts.isCallExpression(node)) {
        if (keywords.has(node.expression.getText())) {
          return ts.createVoidZero();
        }

        const token = node.expression.getFirstToken();
        if (token && keywords.has(token.getText())) {
          return ts.createVoidZero();
        }

        return node;
      }

      return ts.visitEachChild(node, visit, context);
    };
  }

  return function transformer(context: ts.TransformationContext) {
    return (sourceFile: ts.SourceFile) =>
      ts.visitNode(sourceFile, visitor(context, sourceFile));
  };
}
