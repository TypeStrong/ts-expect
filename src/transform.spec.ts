import { describe, it, expect } from "@jest/globals";
import ts from "typescript";
import transform from "./transform";
import { readFileSync } from "fs";

describe("transform", () => {
  const paths = [
    new URL("__fixtures__/test-var-shadow.ts", import.meta.url),
    new URL("__fixtures__/test-import-spread.ts", import.meta.url),
    new URL("__fixtures__/test-import-star.ts", import.meta.url),
  ];

  it("should strip expects", () => {
    const host = ts.createCompilerHost({});
    const program = ts.createProgram(
      paths.map((x) => x.pathname),
      {},
      host
    );
    const result = program.emit(undefined, undefined, undefined, undefined, {
      before: [transform()],
    });

    expect(result.diagnostics).toEqual([]);

    for (const path of paths) {
      const outPath = path.pathname.replace(/\.ts$/, ".js");

      expect(readFileSync(outPath, "utf8")).toMatchSnapshot();
    }
  });
});
