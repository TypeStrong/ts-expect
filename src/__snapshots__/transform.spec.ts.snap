// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform should strip expects 1`] = `
"\\"use strict\\";
exports.__esModule = true;
var ts_expect_1 = require(\\"ts-expect\\");
function test(expectType) {
    return expectType();
}
var test2 = function (expectType) { return expectType(); };
function test3(noShadow) {
    ;
    return noShadow;
}
test(ts_expect_1.expectType);
test2(function () { return undefined; });
test3(123);
;
var expectedType = ts_expect_1.expectType;
expectedType(\\"\\");
"
`;

exports[`transform should strip expects 2`] = `
"\\"use strict\\";
exports.__esModule = true;
var ts_expect_1 = require(\\"ts-expect\\");
;
"
`;

exports[`transform should strip expects 3`] = `
"\\"use strict\\";
exports.__esModule = true;
var expect = require(\\"ts-expect\\");
;
"
`;
